File: OpenBurnBarCore/Sources/OpenBurnBarIrohRelay/IrohRelayFrameCodec.swift
import Foundation
import OpenBurnBarCore

/// Length-prefixed JSON codec for `HermesRealtimeRelayFrame`. The Rust crate
/// writes `[u32 big-endian length][JSON payload]` onto the iroh QUIC stream;
/// this Swift type reads/writes the exact same byte layout so the wire format
/// is identical across transports.
public struct IrohRelayFrameCodec: Sendable {
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder
    private let maxFrameBytes: Int

    public init(maxFrameBytes: Int = IrohRelayProtocol.maxFrameBytes) {
        let encoder = JSONEncoder()
        encoder.outputFormatting = []
        // Drop nils so the on-wire shape matches the Cloud Run relay's
        // `serializeFrame` helper which never emits keys with `undefined`.
        encoder.outputFormatting = [.withoutEscapingSlashes]
        self.encoder = encoder
        self.decoder = JSONDecoder()
        self.maxFrameBytes = maxFrameBytes
    }

    /// Encode a frame into the wire envelope. Throws if the encoded payload
    /// would exceed `maxFrameBytes`.
    public func encode(_ frame: HermesRealtimeRelayFrame) throws -> Data {
        let payload = try encoder.encode(frame)
        guard payload.count <= maxFrameBytes else {
            throw IrohRelayTransportError.streamRejected(
                "iroh relay frame is \(payload.count) bytes, exceeds \(maxFrameBytes)."
            )
        }
        var envelope = Data(capacity: payload.count + IrohRelayProtocol.WireFormat.lengthPrefixBytes)
        var lengthBE = UInt32(payload.count).bigEndian
        withUnsafeBytes(of: &lengthBE) { envelope.append(contentsOf: $0) }
        envelope.append(payload)
        return envelope
    }

    /// Decode an inbound envelope. Surface returns the payload byte range we
    /// consumed so the caller can drain a buffered transport that may have
    /// concatenated several frames. Always re-bases the input view to start
    /// index 0 because `Data.removeFirst(_:)` does not normalize indices in
    /// Foundation, and callers like the loopback transport mutate the buffer
    /// in place between frames.
    public func decode(from envelope: Data) throws -> (frame: HermesRealtimeRelayFrame, consumed: Int) {
        let lengthPrefix = IrohRelayProtocol.WireFormat.lengthPrefixBytes
        guard envelope.count >= lengthPrefix else {
            throw IrohRelayTransportError.decodeFailed("iroh frame envelope is shorter than length prefix.")
        }
        let normalized: Data = envelope.withUnsafeBytes { buffer -> Data in
            Data(buffer)
        }
        var lengthBE: UInt32 = 0
        _ = withUnsafeMutableBytes(of: &lengthBE) { dest in
            normalized.copyBytes(to: dest, from: 0..<lengthPrefix)
        }
        let length = Int(UInt32(bigEndian: lengthBE))
        guard length <= maxFrameBytes else {
            throw IrohRelayTransportError.streamRejected(
                "iroh relay inbound frame is \(length) bytes, exceeds \(maxFrameBytes)."
            )
        }
        let totalBytes = lengthPrefix + length
        guard normalized.count >= totalBytes else {
            throw IrohRelayTransportError.decodeFailed("iroh relay envelope is truncated.")
        }
        let payload = normalized.subdata(in: lengthPrefix..<totalBytes)
        do {
            let frame = try decoder.decode(HermesRealtimeRelayFrame.self, from: payload)
            return (frame, totalBytes)
        } catch {
            throw IrohRelayTransportError.decodeFailed(error.localizedDescription)
        }
    }
}

File: android/openburnbar-iroh-relay/src/main/java/com/openburnbar/irohrelay/IrohRelayFrameCodec.kt
package com.openburnbar.irohrelay

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Length-prefixed JSON codec for `HermesRealtimeRelayFrame`. The Rust
 * crate writes `[u32 big-endian length][JSON payload]` onto the iroh
 * QUIC stream; this class reads/writes the same byte layout so the wire
 * format is byte-identical across iOS, Mac, and Android.
 */
class IrohRelayFrameCodec(
    private val maxFrameBytes: Int = IrohRelayProtocol.MAX_FRAME_BYTES,
) {
    /**
     * Encode a frame into the wire envelope. Throws if the encoded
     * payload would exceed `maxFrameBytes`.
     */
    fun encode(frame: HermesRealtimeRelayFrame): ByteArray {
        val payload = HermesRealtimeRelayJson
            .encodeToString(HermesRealtimeRelayFrame.serializer(), frame)
            .toByteArray(Charsets.UTF_8)
        if (payload.size > maxFrameBytes) {
            throw IrohRelayTransportError.StreamRejected(
                "iroh relay frame is ${payload.size} bytes, exceeds $maxFrameBytes."
            )
        }
        val envelope = ByteArray(payload.size + IrohRelayProtocol.WireFormat.LENGTH_PREFIX_BYTES)
        ByteBuffer.wrap(envelope, 0, IrohRelayProtocol.WireFormat.LENGTH_PREFIX_BYTES)
            .order(ByteOrder.BIG_ENDIAN)
            .putInt(payload.size)
        System.arraycopy(payload, 0, envelope, IrohRelayProtocol.WireFormat.LENGTH_PREFIX_BYTES, payload.size)
        return envelope
    }

    /**
     * Decode an inbound envelope. Returns the parsed frame and the number
     * of bytes consumed from the input.
     */
    fun decode(envelope: ByteArray): DecodedFrame {
        val prefix = IrohRelayProtocol.WireFormat.LENGTH_PREFIX_BYTES
        if (envelope.size < prefix) {
            throw IrohRelayTransportError.DecodeFailed("iroh frame envelope is shorter than length prefix.")
        }
        val length = ByteBuffer.wrap(envelope, 0, prefix).order(ByteOrder.BIG_ENDIAN).int
        if (length < 0 || length > maxFrameBytes) {
            throw IrohRelayTransportError.StreamRejected(
                "iroh relay inbound frame is $length bytes, exceeds $maxFrameBytes."
            )
        }
        val totalBytes = prefix + length
        if (envelope.size < totalBytes) {
            throw IrohRelayTransportError.DecodeFailed("iroh relay envelope is truncated.")
        }
        val payload = envelope.copyOfRange(prefix, totalBytes)
        return try {
            val frame = HermesRealtimeRelayJson.decodeFromString(
                HermesRealtimeRelayFrame.serializer(),
                payload.toString(Charsets.UTF_8),
            )
            DecodedFrame(frame, totalBytes)
        } catch (t: Throwable) {
            throw IrohRelayTransportError.DecodeFailed(t.message ?: t.javaClass.simpleName)
        }
    }

    data class DecodedFrame(val frame: HermesRealtimeRelayFrame, val consumed: Int)
}
